proposals/20230530-driver-kernel-testing-framework.md

```diff
@@ -0,0 +1,454 @@
+
+Furthermore, considering real-world workloads and settings is essential for comprehensive testing and accurate performance evaluation, serving as the guiding principle throughout the process.
+
+
+## Proposal
+
+The "Functionality" tests of the CI-powered test framework are targeted for Falco release 0.36.
+
+- Objective: Ensure Falco drivers run and work across different kernel versions and distributions.
+- Actions: Implement a comprehensive testing process and kernel test grid to validate driver compatibility, stability, and functionality across various kernel environments.
+- Expected Outcome: Increased confidence in the reliability and stability of Falco drivers.
+
+### Test Category 1
+
+Ensuring that the kernel driver successfully compiles for the agreed-upon kernel test grid. The optimal compiler version and build container are carefully selected to align with the advancements in the Linux source tree and its related dependencies. This selection takes into account various factors, including the specific kernel version, driver type, and the distribution being used.
+
+
+### Test Category 2
+
+Verifying that the kernel driver can load, run, and capture events without errors. This is determined through [scap-open](https://github.com/falcosecurity/libs/tree/master/userspace/libscap/examples/01-open) and unit tests conducted in virtual machine (VM) environments. In essence, when we mention that the "driver loads and runs", it implies that the scap-open counter for captured events during a test run is positive and that the [drivers_test](https://github.com/falcosecurity/libs/tree/master/test/drivers) unit tests pass. The latter tests not only load the driver live but also simulate syscall events and verify that the expected information is extracted from the kernel tracepoint and retrieved by the libscap driver type-specific engine in userspace.
+
+
+### Test Infrastructure
+
+Our goal is to facilitate the expansion and integration of full Continuous Integration (CI) for the kernel test grid, enabling thorough testing of the functionality of the kernel drivers. 
+
+Concurrently, we will develop a test framework utilizing localhost virtual machines (VMs), which will include a limited kernel test grid. This framework will be accessible to adopters, providing them with a convenient option for local testing during the development or testing phases.
+
+
+### Current State
+
+Currently (as of May 31, 2023), the relevant [falcosecurity/libs](https://github.com/falcosecurity/libs) CI drivers tests include:
+
+<details>
+    <summary>Build kernel drivers</summary>
+		<ul>
+			<li>Latest Archlinux kernel to spot possible incompatibilities with the latest kernel tree changes</li>
+			<li>ubuntu-22.04 (x86_64)</li>
+			<li>linux-2.6.32</li>
+			<li>linux-3.10</li>
+			<li>linux-4.18</li>
+			<li>linux-5.19</li>
+			<li>linux-6.2</li>
```

Comment:
All of these are on x86_64. 
I am looking into adding same tests for arm64 too.